include/json/config.h

```diff
@@ -67,21 +67,40 @@
 // Storages, and 64 bits integer support is disabled.
 // #define JSON_NO_INT64 1
 
-#if defined(_MSC_VER) && _MSC_VER <= 1200 // MSVC 6
-// Microsoft Visual Studio 6 only support conversion from __int64 to double
-// (no conversion from unsigned __int64).
-#define JSON_USE_INT64_DOUBLE_CONVERSION 1
-// Disable warning 4786 for VS6 caused by STL (identifier was truncated to '255'
-// characters in the debug information)
-// All projects I've ever seen with VS6 were using this globally (not bothering
-// with pragma push/pop).
-#pragma warning(disable : 4786)
-#endif // if defined(_MSC_VER)  &&  _MSC_VER < 1200 // MSVC 6
-
-#if defined(_MSC_VER) && _MSC_VER >= 1500 // MSVC 2008
-/// Indicates that the following function is deprecated.
-#define JSONCPP_DEPRECATED(message) __declspec(deprecated(message))
-#endif
+#if defined(_MSC_VER) // MSVC
+	#if _MSC_VER <= 1200 // MSVC 6
+		// Microsoft Visual Studio 6 only support conversion from __int64 to double
+		// (no conversion from unsigned __int64).
+		#define JSON_USE_INT64_DOUBLE_CONVERSION 1
+		// Disable warning 4786 for VS6 caused by STL (identifier was truncated to '255'
+		// characters in the debug information)
+		// All projects I've ever seen with VS6 were using this globally (not bothering
+		// with pragma push/pop).
+		#pragma warning(disable : 4786)
+	#endif // MSVC 6
+
+	#if _MSC_VER >= 1500 // MSVC 2008
+		/// Indicates that the following function is deprecated.
+		#define JSONCPP_DEPRECATED(message) __declspec(deprecated(message))
+	#endif
+
+	#if _MSC_VER >= 1600 // MSVC 2010
+		#define JSON_HAS_RVALUE_REFERENCES
+	#endif 
+#endif // defined(_MSC_VER)
+
+
+#ifdef __clang__
+	#if __has_feature(cxx_rvalue_references)
+		#define JSON_HAS_RVALUE_REFERENCES
+	#endif 
+#elif defined __GNUC__
+	#if defined(__GXX_EXPERIMENTAL_CXX0X__) 
```

Comment:
That is not exactly how boost is doing it.
Boost.config is doing this for GCC:

```c++
#if defined(__GXX_EXPERIMENTAL_CXX0X__) || (__cplusplus >= 201103L)
#  define BOOST_GCC_CXX11
#endif
```

http://www.boost.org/doc/libs/1_58_0/boost/config/compiler/gcc.hpp

That's just gcc.hpp, but each compiler is different.

This is the built-in compiler support. Boost can also be reconfigured for unforeseen compilers in which case it tries a little test program to see if rvalue references are supported.

The analogous macros for VisualC++ are here:
http://www.boost.org/doc/libs/1_58_0/boost/config/compiler/visualc.hpp